### PR TITLE
tests for kafka webhook were implemented & clusters webhooks were refactored

### PR DESCRIPTION
**Pull request by tengu-alt**
Covered Kafka webhook with tests and fixed bug in clusters datacentre name updating validation flow

### CODE REVIEW

File: apis/clusters/v1beta1/kafka_webhook_test.go [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+			Expect(k8sClient.Create(ctx, &TestKafkaManifest)).ShouldNot(Succeed())
+
+			TestKafkaManifest.Spec.Kraft = []*Kraft{TestKafkaManifest.Spec.Kraft[0]}
+			TestKafkaManifest.Spec.Kraft[0].ControllerNodeCount = 4
+			Expect(k8sClient.Create(ctx, &TestKafkaManifest)).ShouldNot(Succeed())
+			TestKafkaManifest.Spec.Kraft[0].ControllerNodeCount = 3
+
+			done := NewChannelWithTimeout(timeout)
+
+			<-done
```
Comment (worryg0d) [inline apis/clusters/v1beta1/kafka_webhook_test.go:66]: It looks unnecessary, what is the point?

File: apis/clusters/v1beta1/kafka_webhook_test.go [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+
+			prevCloudProviderSettings := kafkaManifest.Spec.DataCentres[0].CloudProviderSettings
+			TestKafkaManifest.Spec.DataCentres[0].CloudProviderSettings = []*CloudProviderSettings{prevCloudProviderSettings[0], prevCloudProviderSettings[0]}
+			Expect(k8sClient.Patch(ctx, &TestKafkaManifest, patch)).ShouldNot(Succeed())
+			TestKafkaManifest.Spec.DataCentres[0].CloudProviderSettings = prevCloudProviderSettings
+
+			prevMapField := TestKafkaManifest.Spec.DataCentres[0].Tags
+			TestKafkaManifest.Spec.DataCentres[0].Tags["test"] = "test"
+			Expect(k8sClient.Patch(ctx, &TestKafkaManifest, patch)).ShouldNot(Succeed())
+			TestKafkaManifest.Spec.DataCentres[0].Tags = prevMapField
```
Comment (worryg0d) [inline apis/clusters/v1beta1/kafka_webhook_test.go:221]: Map is a reference type so they should be identical when you returns the previous map state. You should create a new map and assign it to the test kafka manifest.

File: apis/clusters/v1beta1/kafka_webhook_test.go (hunk 1 above)
Comment (tengu-alt) [reply apis/clusters/v1beta1/kafka_webhook_test.go]: agree, removed

File: apis/clusters/v1beta1/kafka_webhook_test.go (hunk 2 above)
Comment (tengu-alt) [reply apis/clusters/v1beta1/kafka_webhook_test.go]: agree, and added other cases

File: apis/clusters/v1beta1/kafka_webhook_test.go [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+
+			if len(TestKafkaManifest.Spec.TwoFactorDelete) == 0 {
+				TestKafkaManifest.Spec.TwoFactorDelete = []*TwoFactorDelete{{Email: "test@test.com"}, {Email: "test@test.com"}}
+			} else {
+				TestKafkaManifest.Spec.TwoFactorDelete = []*TwoFactorDelete{kafkaManifest.Spec.TwoFactorDelete[0], {Email: "test@test.com"}}
+			}
+			Expect(k8sClient.Create(ctx, &TestKafkaManifest)).ShouldNot(Succeed())
+			TestKafkaManifest.Spec.TwoFactorDelete = kafkaManifest.Spec.TwoFactorDelete
+
+			TestKafkaManifest.Spec.SLATier = "some SLATier that is not support"
```
Comment (testisnullus) [inline apis/clusters/v1beta1/kafka_webhook_test.go:48]: some SLATier that is not support -> some SLATier that is not supported

File: apis/clusters/v1beta1/kafka_webhook_test.go [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+	})
+
+	When("updating a Kafka manifest", func() {
+		It("should test kafka update flow", func() {
+			TestKafkaManifest.Status.State = models.RunningStatus
+			Expect(k8sClient.Create(ctx, &TestKafkaManifest)).Should(Succeed())
+
+			patch := TestKafkaManifest.NewPatch()
+			TestKafkaManifest.Status.ID = models.CreatedEvent
+			Expect(k8sClient.Status().Patch(ctx, &TestKafkaManifest, patch)).Should(Succeed())
```
Comment (testisnullus) [inline apis/clusters/v1beta1/kafka_webhook_test.go:73]: Why do we set the `models.CreatedEvent` to the status ID?

File: apis/clusters/v1beta1/kafka_webhook_test.go [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+
+			prevSchemaRegistry := kafkaManifest.Spec.SchemaRegistry
+			TestKafkaManifest.Spec.SchemaRegistry = []*SchemaRegistry{prevSchemaRegistry[0], prevSchemaRegistry[0]}
+			Expect(k8sClient.Patch(ctx, &TestKafkaManifest, patch)).ShouldNot(Succeed())
+			TestKafkaManifest.Spec.SchemaRegistry = prevSchemaRegistry
+
+			prevKarapaceSchemaRegistry := kafkaManifest.Spec.KarapaceSchemaRegistry
+			TestKafkaManifest.Spec.KarapaceSchemaRegistry = []*KarapaceSchemaRegistry{prevKarapaceSchemaRegistry[0], prevKarapaceSchemaRegistry[0]}
+			Expect(k8sClient.Patch(ctx, &TestKafkaManifest, patch)).ShouldNot(Succeed())
+			TestKafkaManifest.Spec.KarapaceSchemaRegistry = prevKarapaceSchemaRegistry
```
Comment (testisnullus) [inline apis/clusters/v1beta1/kafka_webhook_test.go:127]: Should we also test that the values itself were not changed?

File: apis/clusters/v1beta1/kafka_webhook_test.go (hunk 4 above)
Comment (tengu-alt) [reply apis/clusters/v1beta1/kafka_webhook_test.go:73]: discussed

File: apis/clusters/v1beta1/kafka_webhook_test.go (hunk 5 above)
Comment (tengu-alt) [reply apis/clusters/v1beta1/kafka_webhook_test.go:127]: discussed

File: apis/clusters/v1beta1/kafka_webhook_test.go (hunk 3 above)
Comment (tengu-alt) [reply apis/clusters/v1beta1/kafka_webhook_test.go:48]: agree, changed

**Review by worryg0d — APPROVED** [review]

File: apis/clusters/v1beta1/kafka_webhook_test.go [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/yaml"
+
+	"github.com/instaclustr/operator/pkg/models"
+)
+
+var _ = Describe("Kafka Controller", Ordered, func() {
+	kafkaManifest := Kafka{}
```
Comment (ribaraka) [inline apis/clusters/v1beta1/kafka_webhook_test.go:16]: `validKafka`

**Review by ribaraka — CHANGES_REQUESTED** [review]

File: apis/clusters/v1beta1/kafka_webhook_test.go [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+	TestKafkaManifest := Kafka{}
+
+	It("Reading kafka manifest", func() {
+		yfile, err := os.ReadFile("../../../controllers/clusters/datatest/kafka_v1beta1.yaml")
+		Expect(err).Should(Succeed())
+
+		err = yaml.Unmarshal(yfile, &kafkaManifest)
+		Expect(err).Should(Succeed())
+		kafkaNamespacedName.Name = kafkaManifest.ObjectMeta.Name
+		kafkaNamespacedName.Namespace = defaultNS
```
[the remark this answers is not in the document]
Comment (tengu-alt) [reply apis/clusters/v1beta1/kafka_webhook_test.go:27]: agree, fixed

File: apis/clusters/v1beta1/kafka_webhook_test.go [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/yaml"
+
+	"github.com/instaclustr/operator/pkg/models"
+)
+
+var _ = Describe("Kafka Controller", Ordered, func() {
+	kafkaManifest := Kafka{}
+	kafkaNamespacedName := &types.NamespacedName{}
+	TestKafkaManifest := Kafka{}
```
[the remark this answers is not in the document]
Comment (tengu-alt) [reply apis/clusters/v1beta1/kafka_webhook_test.go:18]: IMHO TestKafkaManifest is more appropriate

File: apis/clusters/v1beta1/kafka_webhook_test.go [hunk 9]
```diff
@@ -0,0 +1,227 @@
...
+
+			TestKafkaManifest.Spec.Description = "some description"
+			Expect(k8sClient.Create(ctx, &TestKafkaManifest)).ShouldNot(Succeed())
+			TestKafkaManifest.Spec.Description = kafkaManifest.Spec.Description
+
+			if len(TestKafkaManifest.Spec.TwoFactorDelete) == 0 {
+				TestKafkaManifest.Spec.TwoFactorDelete = []*TwoFactorDelete{{Email: "test@test.com"}, {Email: "test@test.com"}}
+			} else {
+				TestKafkaManifest.Spec.TwoFactorDelete = []*TwoFactorDelete{kafkaManifest.Spec.TwoFactorDelete[0], {Email: "test@test.com"}}
+			}
```
[the remark this answers is not in the document]
Comment (tengu-alt) [reply apis/clusters/v1beta1/kafka_webhook_test.go:44]: agree, fixed

File: apis/clusters/v1beta1/kafka_webhook_test.go [hunk 10]
```diff
@@ -0,0 +1,227 @@
...
+				TestKafkaManifest.Spec.Kraft = []*Kraft{kafkaManifest.Spec.Kraft[0], {ControllerNodeCount: 1}}
+			}
+			Expect(k8sClient.Create(ctx, &TestKafkaManifest)).ShouldNot(Succeed())
+
+			TestKafkaManifest.Spec.Kraft = []*Kraft{TestKafkaManifest.Spec.Kraft[0]}
+			TestKafkaManifest.Spec.Kraft[0].ControllerNodeCount = 4
+			Expect(k8sClient.Create(ctx, &TestKafkaManifest)).ShouldNot(Succeed())
+			TestKafkaManifest.Spec.Kraft[0].ControllerNodeCount = 3
+		})
+	})
```
[the remark this answers is not in the document]
Comment (tengu-alt) [reply apis/clusters/v1beta1/kafka_webhook_test.go:64]: it's setting previous valid value

File: apis/clusters/v1beta1/kafka_webhook_test.go (hunk 6 above)
Comment (tengu-alt) [reply apis/clusters/v1beta1/kafka_webhook_test.go:16]: also as with Testkafkamanifest

**Review by ribaraka — APPROVED** [review]

**Pull request merged**